crypto/box/src/hkdf_box.rs

```diff
@@ -0,0 +1,167 @@
+    Sum<<KexAlgo::Public as PublicKey>::Size, AeadAlgo::TagSize>: ArrayLength<u8>,
+{
+    type FooterSize = Sum<<KexAlgo::Public as PublicKey>::Size, AeadAlgo::TagSize>;
+
+    fn encrypt_in_place_detached<T: RngCore + CryptoRng>(
+        &self,
+        rng: &mut T,
+        key: &KexAlgo::Public,
+        buffer: &mut [u8],
+    ) -> Result<GenericArray<u8, Self::FooterSize>, AeadError> {
+        // ECDH
+        use keys::KexPublic;
+        let (our_public, shared_secret) = key.new_secret(rng);
+
+        let curve_point_bytes =
+            GenericArray::<u8, <KexAlgo::Public as PublicKey>::Size>::clone_from_slice(
+                our_public.get_bytes().as_ref(),
+            );
+
+        // KDF
+        let (aes_key, aes_nonce) = Self::kdf_step(&shared_secret);
+
+        // AES
+        let aead = AeadAlgo::new(aes_key);
+        let mac = aead.encrypt_in_place_detached(&aes_nonce, &[], buffer)?;
+
+        // Tag is curve_point_bytes || aes_mac_bytes
+        Ok(curve_point_bytes.concat(mac))
+    }
+
+    fn decrypt_in_place_detached(
+        &self,
+        key: &KexAlgo::Private,
+        tag: &GenericArray<u8, Self::FooterSize>,
+        buffer: &mut [u8],
+    ) -> Result<(), Error> {
+        // ECDH
+        use keys::KexReusablePrivate;
+        let public_key =
+            KexAlgo::Public::try_from(&tag[..<KexAlgo::Public as PublicKey>::Size::USIZE])
+                .map_err(Error::Key)?;
+        let shared_secret = key.key_exchange(&public_key);
+
+        // KDF
+        let (aes_key, aes_nonce) = Self::kdf_step(&shared_secret);
+
+        // AES
+        let mac_ref = <&GenericArray<u8, AeadAlgo::TagSize>>::from(
+            &tag[<KexAlgo::Public as PublicKey>::Size::USIZE..],
+        );
+        let aead = AeadAlgo::new(aes_key);
+        aead.decrypt_in_place_detached(&aes_nonce, &[], buffer, mac_ref)
+            .map_err(|_| Error::MacFailed)?;
+
+        Ok(())
+    }
+}
+
+impl<KexAlgo, DigestAlgo, AeadAlgo> HkdfBox<KexAlgo, DigestAlgo, AeadAlgo>
+where
+    KexAlgo: Kex,
+    for<'privkey> <KexAlgo as Kex>::Public: From<&'privkey <KexAlgo as Kex>::EphemeralPrivate>,
+    DigestAlgo: Digest + Input + FixedOutput + Default + Clone + BlockInput + Reset,
+    AeadAlgo: Aead + NewAead,
+    AeadAlgo::NonceSize:
+        IsLessOrEqual<<KexAlgo::Secret as KexSecret>::EntropyLowerBound, Output = B1>,
+    AeadAlgo::KeySize:
+        IsLessOrEqual<<KexAlgo::Secret as KexSecret>::EntropyLowerBound, Output = B1>,
+    <<KexAlgo as Kex>::Public as PublicKey>::Size: Add<<AeadAlgo as aead::Aead>::TagSize>,
+    Sum<<KexAlgo::Public as PublicKey>::Size, AeadAlgo::TagSize>: ArrayLength<u8>,
+{
+    /// KDF part, factored out to avoid duplication
+    /// This part must produce the key and IV/nonce for aes-gcm
+    /// Blake2b produces 64 bytes of private key material which is more than we need,
+    /// so we don't do the HKDF-EXPAND step.
```

Comment:
I have to update this comment. Until rust implements Specialization of generics, there's not an easy way to implement that optimization. So we now just always do the HKDF-EXPAND step.